### ISSUE TITLE
[2022/03/10]推薦・機械学習勉強会

### ISSUE THREAD

**Issue by yu-ya4** [wantedly/machine-learning-round-table#134]
## Why

推薦・機械学習勉強会は、推薦や機械学習、その周辺技術を通じてサービスを改善することにモチベーションのある人達の集まりです。ニュースやブログから論文まで、気になったものについてお互い共有しましょう！

発信のため、ここは **public** にしてあります。外部からの参加をご希望の方は松村(https://twitter.com/yu__ya4) まで DM を送るか、Wantedly Visit の募集（https://www.wantedly.com/projects/391912） よりご連絡ください！

## What

Wantedly では隔週木曜日に

- 推薦の評価指標について議論したい
- 〇〇っていうライブラリ / フレームワークを導入してみたい
- 他社の基盤事例をみんなにシェアして自社の基盤開発に活かしたい
- もっと推薦を良くするためにどんなものが必要か議論したい

といった話をする「推薦・機械学習勉強会」を開催しています。
この ISSUE はその会で話すネタを共有するための場所です。

話したいことがある人はここにコメントしましょう！
会の間に話した内容もここにメモしましょう！

prev: #132

**Comment by Hayashi-Yudai** [comment]
# [Hierarchical representations improve image retrieval](https://www.amazon.science/blog/hierarchical-representations-improve-image-retrieval)

Amzonでは画像から商品が検索できるシステムがある。そこで画像検索の精度を向上させるために階層構造をモデルに取り込んだ。

<img width="429" alt="image" src="https://user-images.githubusercontent.com/34836226/157630970-1d89aa08-fd2f-43e0-a638-5a18773b37f4.png">

Deep Metric Learning (DML) で画像のembeddingと距離関数を学習する。似通っている画像のembedding間の距離が学習が進められるが、ここに階層構造を導入することで、同じsuperclassを持つクラスの画像同士は距離が近くなるように学習できる。

<img width="903" alt="image" src="https://user-images.githubusercontent.com/34836226/157631095-424ce7a6-9e32-4388-9d29-af63b77aa779.png">

これによって未知のデータへ適用したときにも良い結果が出る。

階層構造が肘名なときにはk-meansでクラスタリングして雑に階層構造を作っておいて学習のステップでそれを精製していく。

# [Push Notification Best Practices: 35 Tips for Dramatically Better Messages](https://clevertap.com/blog/push-notification-best-practices/)

プッシュ通知のベストプラクティス集。ここから逆算してデータ分析に活かせそう。

**Comment by zerebom** [comment]
## [エンジニアも知っておきたいAI倫理のはなし](https://www.slideshare.net/ssuserf7b6ad/ai-251212432)

- AI倫理は誰が議論してる?
    - 各国政府や企業
- AIエンジニアが倫理を学ぶ必要性
    - 意図せず人を傷つけるのを防ぐ
    - 自分のバイアスに気づく
- 過去の事例
    - Google: が写真のタグ付けで人に誤ったラベルを付与した
    - 警察: 予測操作
    - Amazon: 男性を高く評価するシステムが原因でPJチームが解散
    - リクナビ: 内定辞退率
- 世の動向
    - 2017: 議論が開始 → 2022: AI論理の原則をどのように社会実装するか
    - 多くの団体がAI原則を公表
- エンジニアが倫理インシデントを起こさないためには
    - やばいと感じたら声をあげる
    - 不十分なことによるリスクを認識する
    - 他の会社ガイドラインを見てみる


## [データサイエンティストがオブジェクト指向を学んでみた際のメモ〜scikit-learnの作法を知る](http://kamonohashiperry.com/archives/5136)
DSのカモノハシさんがオブジェクト指向を学んだ時のメモ。
ブログの前半でオブジェクト指向の概要、後半で: sklearnのコードリーディングを紹介している。

例えばKMeansは以下のクラスを継承している。
  
  ```
  _ClassNamePrefixFeaturesOutMixin：特徴量の名前を取得するためのもの
  TransformerMixin：データの変換のためのもの
  ClusterMixin：クラスタリングでラベルを返すためのもの
  BaseEstimator：あらゆる推定におけるベースとなるもの
  ```

<details>
<summary>RegressorMixin</summary>

```python
class RegressorMixin:
    """Mixin class for all regression estimators in scikit-learn."""

    _estimator_type = "regressor"

    def score(self, X, y, sample_weight=None):
        """Return the coefficient of determination of the prediction.

        The coefficient of determination :math:`R^2` is defined as
        :math:`(1 - \\frac{u}{v})`, where :math:`u` is the residual
        sum of squares ``((y_true - y_pred)** 2).sum()`` and :math:`v`
        is the total sum of squares ``((y_true - y_true.mean()) ** 2).sum()``.
        The best possible score is 1.0 and it can be negative (because the
        model can be arbitrarily worse). A constant model that always predicts
        the expected value of `y`, disregarding the input features, would get
        a :math:`R^2` score of 0.0.

        Parameters
        ----------
        X : array-like of shape (n_samples, n_features)
            Test samples. For some estimators this may be a precomputed
            kernel matrix or a list of generic objects instead with shape
            ``(n_samples, n_samples_fitted)``, where ``n_samples_fitted``
            is the number of samples used in the fitting for the estimator.

        y : array-like of shape (n_samples,) or (n_samples, n_outputs)
            True values for `X`.

        sample_weight : array-like of shape (n_samples,), default=None
            Sample weights.

        Returns
        -------
        score : float
            :math:`R^2` of ``self.predict(X)`` wrt. `y`.

        Notes
        -----
        The :math:`R^2` score used when calling ``score`` on a regressor uses
        ``multioutput='uniform_average'`` from version 0.23 to keep consistent
        with default value of :func:`~sklearn.metrics.r2_score`.
        This influences the ``score`` method of all the multioutput
        regressors (except for
        :class:`~sklearn.multioutput.MultiOutputRegressor`).
        """

        from .metrics import r2_score

        y_pred = self.predict(X)
        return r2_score(y, y_pred, sample_weight=sample_weight)

    def _more_tags(self):
        return {"requires_y": True}
```

</details>


## [新卒データサイエンティスト研修「DSOps研修」とは何か?](https://developers.cyberagent.co.jp/blog/archives/34628/)

「データサイエンティストがビジネス価値を出すために，そもそも何を知っていないといけないか?」という問いに根ざした研修。受けたい。

**Comment by yu-ya4** [comment]
## より良いデータ分析

『Google が公開している、より良いデータ分析のためのガイドブック「Good Data Analysis」で、データ分析の要所が簡潔にまとめられていて感動した』

https://shunyaueta.com/posts/2022-03-08/

> Data analysis starts with questions, not data or a technique
> データ分析は質問とともにのみ始まる、データや技術からデータ分析は始まらない

とりあえず分析()するってのはだいたい無駄ﾜｶﾙ

## RecSys2022 Challenge が始まりましたね

http://www.recsyschallenge.com/2022/

>This year’s challenge focuses on fashion recommendations. When given user sessions, purchase data and content data about items, can you accurately predict which fashion item will be bought at the end of the session?

ファッション👕
